### ISSUE TITLE
[RFC] [new feature] API for custom lattices

### ISSUE THREAD

**Issue by attila-i-szabo** [netket/netket#1049]
The main limitation of `Lattice` is that we can only colour edges by the distance between the sites. This is restrictive, e.g., we couldn't implement a Kitaev honeycomb model without defining an alternative `Graph`.

I plan to add support for more general lattices soon. Basically, you can specify any configuration of bonds in a single unit cell and it can be repeated for the whole lattice automatically. The question is how the user should provide this data.

1. For each bond, we need the following information:
    * which atom in the unit cell you start from
    * which atom in the unit cell you go to
    * if you need to move into a different unit cell
    * the bond colour
  
    In the input, I would keep all of this together for one bond, and then go to the next, so it would be like
    ```python
    bonds = [
        (from, to, [Δx, Δy], col),
        # next bond
    ]
    ```
    The alternative would be keeping the different kinds of data separate, so we'd pass an array of `from`, `to`, `Δr`, `col` to the constructor, like `bondops` and `bondops_colors` in `GraphOperator`. I don't like this because it would be much harder to read what belongs together across 3-4 distinct parameters.

2. We can either specify `Δr` either in terms of the number of unit cells you have to move or as a Cartesian displacement vector. Both have their advantages and it is easy to convert from one to the other inside the constructor. I prefer the Cartesian option because I had to define some lattices with unconventional unit cells, which makes it hard to work out Δ(unit cell) by hand.

I'd be interested to hear what you think about these, it's easier to change the API before you write everything!

@femtobit @PhilipVinc @VolodyaCO

**Comment by gcarleo** [comment]
@attila-i-szabo, this sounds like an interesting proposal but I am just worried that this is going to be a very specific use case (say, Kitaev honeycomb). Can you maybe clarify if you can think of other more general use cases where this situation could happen and that are not supported by the regular lattice? Is this kind of functionality supported in other graph libraries?

**Comment by femtobit** [comment]
Thanks for writing down your proposal here. Some first thoughts:

1. I agree that this is a desirable feature and the data format you propose for specifying the bonds looks sensible to me and I also like the idea of giving your `Δr` as plain position-space vectors.

2. I also think it would be best to support specifying this kind of bond setup on creation of the lattice instead of passing them only to `GraphOperator`. (The only cleaner solution I see would be to have a full separation between the geometric lattice described by `Lattice` and the graph structure given by the bonds, but as our `Lattice` inherits from `Graph`, this cannot be easily changed -- at least until NetKet 4 :grin: -- and its not a huge issue, so lets not touch that.)

3. This way of specifying bonds seems to be a full; alternative to the next-neighbor bond creation currently done by `Lattice`. Or is there a kind-of mixed mode where users would want both edges from the current neighbor-detection and _additionally_ add bonds via your syntax? If not, i.e., if a graph can be constructed using either the old our new logic that would make implementation a lot simpler, I think.

One way to implement this without interfering with the current code would be to provide an alternative constructor for `Lattice` as a `staticmethod`. For example
```python3
lat = Lattice.with_custom_bonds(
    basis_vectors=...,
    sites=...,
    bonds=[
        (from, to, [Δx, Δy], col),
        ...,
    ],
)
```

**Comment by PhilipVinc** [comment]
> Is this kind of functionality supported in other graph libraries?

The issue at hand, I imagine, is that by now `Lattice` has very little left to do with a graph, and includes a lot of code to automatically generate the point group/space group/symmetries ecc... 

I am a bit worried because graphs and lattices have seen a lot of iteration and new features recently, but those are somewhat orthogonal to netket itself, however we are bound because of stability reasons to support them. And those add some additional weights to future refactors. 

I have a feeling `nk.graph.Lattice` together with all the space group/symmetry stuff is outgrowing netket itself.

Maybe I'm going off topic, and maybe it's not relevant for this particular feature which maybe makes total sense, but we should try to ensure that the underlying netket representation can work with anything else and then allow users to construct lattices however they want, much akin to how LocalOperators accept a list of operators and `acting_on` but can also be built with the nicer interface of adding single operators...

**Comment by femtobit** [comment]
> try to ensure that the underlying netket representation can work with anything else and then allow users to construct lattices however they want,

As a side note, its of course possible for users to color edges themselves using their own logic. If we add support for `Lattice` to set custom edge colors, that would make it completely customizable. (This can already be done in principle by extracting the edges from `Lattice` and creating a custom `Graph` using those edges and custom colors; tt's just not very intuitive.) This would go together with @wdphy16's recent feature request concerning node colors (#1033).

**Comment by PhilipVinc** [comment]
And I think @attila-i-szabo wants to generate the correct symmetry groups for the coloured lattice, am I right?

**Comment by attila-i-szabo** [comment]
I reply to the main points in no particular order.

> Is this kind of functionality supported in other graph libraries?

Not sure what you mean by "graph libraries". For other libraries for physics on lattices, this is the usual way of specifying bonds on a lattice (AFAI can tell, qutip only supports this kind of format).

> Can you maybe clarify if you can think of other more general use cases where this situation could happen and that are not supported by the regular lattice?

I can think of one more example: there are two inequivalent kinds third neighbour on the kagome (across the hexagon and along a line), but they behave quite differently and there is some literature on what phases they can stabilise. 

> but I am just worried that this is going to be a very specific use case (say, Kitaev honeycomb).

Well, having this machinery and not being able to support the Kitaev honeycomb easily does sound like a problem. Also, I don't like to second-guess what users might want to do.

> This way of specifying bonds seems to be a full alternative to the next-neighbor bond creation currently done by `Lattice`. Or is there a kind-of mixed mode where users would want both edges from the current neighbor-detection and _additionally_ add bonds via your syntax? If not, i.e., if a graph can be constructed using either the old our new logic that would make implementation a lot simpler, I think.

Agreed, I think allowing both definition syntaxes in the same call would only invite problems. There might be cases when it's convenient, but more likely users would end up specifying some nearest-neighbour bonds both ways... And in the former cases, you can still define two Lattices and add the resulting `GraphOperator`s later.

> One way to implement this without interfering with the current code would be to provide an alternative constructor for `Lattice` as a `staticmethod`. For example

Why not support it in the constructor? (Other than python not supporting two constructors in the class...) My idea of how to do it was a constructor that contains all the logic for setting the sites and two private methods that can construct the edges using either logic. It would anyway be nice to make the nearest-neighbour logic more modular, I've always had a hard time following what happens where in that constructor

> As a side note, it's of course possible for users to color edges themselves using their own logic. If we add support for `Lattice` to set custom edge colors, that would make it completely customizable. (This can already be done in principle by extracting the edges from `Lattice` and creating a custom `Graph` using those edges and custom colors; tt's just not very intuitive.)

Of course that's true, but it's kinda ridiculous that we have this huge symmetry interface and we can't do something as basic as defining a Kitaev honeycomb without doing this. Also, implementing this is a lot simpler than the existing edge-finding logic, it feels like we're trapped in thinking this is an advanced feature by having written the (much more advanced) near-neighbour search early on...

> This would go together with @wdphy16's recent feature request concerning node colors (#1033).

That's a bit different. The basic idea of `Lattice` is that it's a graph with a repeating unit cell. I want to make the edges in each unit cell fully customisable, which doesn't break this premise. IIUC, #1033 would colour equivalent nodes in different unit cells to different colours, which does. Also, it seems it's more about exposing node colouring in `igraph` in a generic `Graph`, which is a separate issue.

**Comment by attila-i-szabo** [comment]
> The issue at hand, I imagine, is that by now `Lattice` has very little left to do with a graph, and includes a lot of code to automatically generate the point group/space group/symmetries ecc...
> 
> I am a bit worried because graphs and lattices have seen a lot of iteration and new features recently, but those are somewhat orthogonal to netket itself, however we are bound because of stability reasons to support them. And those add some additional weights to future refactors.
> 
> I have a feeling `nk.graph.Lattice` together with all the space group/symmetry stuff is outgrowing netket itself.
> 
> Maybe I'm going off topic, and maybe it's not relevant for this particular feature which maybe makes total sense, but we should try to ensure that the underlying netket representation can work with anything else and then allow users to construct lattices however they want, much akin to how LocalOperators accept a list of operators and `acting_on` but can also be built with the nicer interface of adding single operators...

This is kinda off topic, but probably a discussion worth having in another thread.

I've said already that I view `Lattice` as a graph with a repeating unit cell and geometric information about the sites. The whole symmetry machinery is a consequence of that—you need to know where sites are if you want to apply a point-group symmetry to them. There are two equally sensible ways of supporting this:
* as a `Graph` with extra geometric information
* as a generator of `Graph`s

The second one would have the advantage that you could decouple geometry from the connectivity information (which is all you need for a `GraphOperator`), and generate a number of graphs from the same `Lattice` (e.g., you could yield a graph of first, second,... neighbours without deciding the maximum number in advance, or of a custom set of edges). 

However, we already have a well-functioning interface following the first option, and changing to the second one would break a lot of things. And in any case, it has little to do with supporting such a basic feature.

**Comment by femtobit** [comment]
> Why not support it in the constructor? (Other than python not supporting two constructors in the class...) My idea of how to do it was a constructor that contains all the logic for setting the sites and two private methods that can construct the edges using either logic. It would anyway be nice to make the nearest-neighbour logic more modular, I've always had a hard time following what happens where in that constructor

You do already mention my main reason for suggesting not to do this in the constructor: Python does not support constructor overloading, which is why it is a relatively common pattern to use named static methods like `lattice.with_custom_bonds` or similar instead.

We could argue whether the interface proposed here should have been used for the default constructor or not, but for historical reasons the nearest-neighbor search is what is done in `__init__` now and thus we are bound to continue supporting this in any case. Adding two essentially independent code paths for initialization to the `Lattice` constructor seems overly complicated to me (and would also be harder to explain in the documentation). Hence my suggestion of using a separate method. That's easier to document and backwards-compatible without additional work.

**Comment by attila-i-szabo** [comment]
I had a look at how to make factory methods in python. Unless I overlooked something, a `classmethod` like `Lattice.with_custom_bonds` would also have to call the (only) constructor to get a `Lattice` object. Since we don't want that object to have any autogenerated bonds, we'd still need a new code path in the constructor to get an edgeless lattice. I'm not sure whether it would be much neater to do this than adding extra arguments to the constructor. As for the implementation, I'd separate the both kinds of edge-generation logic into separate functions that are just called in the appropriate place anyway.

**Comment by femtobit** [comment]
> I had a look at how to make factory methods in python. Unless I overlooked something, a classmethod like Lattice.with_custom_bonds would also have to call the (only) constructor to get a Lattice object.

@attila-i-szabo There is a workaround: It is possible to manually create the object using `__new__` in a classmethod (which does not call `__init__`), then initializing and returning it. This is done below (line 93), for example, to implement `Graph.from_igraph`, circumventing `Graph.__init__`:
https://github.com/netket/netket/blob/30d938efe78798dc0ee6da39df12abfa76978b05/netket/graph/graph.py#L86-L105

Since `__init__` is not called, care needs to be taken to fully initialize the object (i.e., not leaving out any instance attributes `__init__` would have added) so that everything works as expected.

**Comment by PhilipVinc** [comment]
I am against using such hacks because if later on we find a bug in the constructor this means we have to fix two code-paths.

In my mind, the `__init__` method should always be executed and contain sufficiently generic code to construct a completely arbitrary object.
If we want to add additional constructors those should be built on top of `__init__` and convert whatever data they are fed to a format that `init` can process.

**Comment by attila-i-szabo** [comment]
Just a quick question to those who know the current implementation well (@femtobit?) - what does `get_true_edges` do? If I understand it correctly, we check meticulously whether any of the edges are duplicated, self-referential, etc. [here](https://github.com/netket/netket/blob/980c2810e6a9470f286dbab48bcacc73015a3b71/netket/graph/lattice.py#L144-L157) but then we run the resulting graphs past igraph's `simplify` method [here](https://github.com/netket/netket/blob/980c2810e6a9470f286dbab48bcacc73015a3b71/netket/graph/lattice.py#L340-L347), which is supposed to the same thing. Is this really just duplicating the same work, or is something more happening? I'd like to know so I don't break anything while refactoring to make room for the new construction path.

**Comment by femtobit** [comment]
Judging from your PR, I take it you figured it out. Indeed the edge creation code had some redundancies. I originally kept most of the older logic when I rewrote the API, since it was working correctly albeit in a somewhat convoluted way.

> question 2: in the first snippet, are all the inequality checks really needed seeing they go into a set?

I agree what you are now doing in the PR (just using the set, but making sure only ordered edges `node1 < node2` are added) is equivalent and more concise. The old checks were left over from when `true_edges` was a `list`, not a `set`. I changed it because using the list was excruciatingly slow, but you are right, I could have simplified the checks at that time too.
